### ISSUE TITLE
Support gradient checkpointing for functional models via the standard fit method

### ISSUE THREAD

**Issue by rivershah** [keras-team/tf-keras#98]
**System information**.

TensorFlow version (you are using): `2.12.0`
Are you willing to contribute it (Yes/No) : No

**Describe the feature and the current behavior/state**.

Describe the feature clearly here. Be sure to convey here why the requested feature is needed. Any brief description about the use-case would help.

The following feature excerpt is copied from `huggingface` documentation [here](https://huggingface.co/docs/transformers/v4.18.0/en/performance#gradient-checkpointing)

In order to compute the gradients during the backward pass all activations from the forward pass are normally saved. This can create a big memory overhead. Alternatively, one could forget all activations during the forward pass and recompute them on demand during the backward pass. This would however add a significant computational overhead and slow down training.

Gradient checkpointing strikes a compromise between the two approaches and saves strategically selected activations throughout the computational graph so only a fraction of the activations need to be re-computed for the gradients. See [this great article](https://medium.com/tensorflow/fitting-larger-networks-into-memory-583e3c758ff9) explaining the ideas behind gradient checkpointing.

**Will this change the current api? How?**

Yes, it will introduce a flag in `model.fit` such as `gradient_checkpointing=True`

**Who will benefit from this feature?**

Any user with large intermediate activations and tensors that won't fit in device memory. Sequence models can be very memory heavy and this allows scale up without introducing model parallelism complexity. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): No
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Comment by innat-asj** [comment]
@sachinprasadhs 
Could you please explain why this issue moved from keras to here? As it is a feature request, and other framework like torch support gradient checkpoint. That, IMO, makes it appropriate feature for keras.

**Comment by sachinprasadhs** [comment]
Hi, This was moved as part of migration of Keras 3, if you feel this issue is related to Keras 3, feel free to open a new issue in Keras repo.

**Comment by innat** [comment]
Thanks for the response. Actually, there are some issue which are moved from keras to tf.keras repo, which doesn't belong to any specific version, rather they are general. 

I would suggest to evaluate those issues and upon that, send them back to the keras repo. Also recreating new issue is not necessary, as issue can be transfered. 

Here are some other issue, please evaluate. 

https://github.com/keras-team/tf-keras/issues/330
https://github.com/keras-team/tf-keras/issues/350

**Comment by sachinprasadhs** [comment]
Thank you for reporting the issue. 
We're now prioritizing [Keras 3](https://keras.io/keras_3/) development, a multi-backend framework that offers significant advantages by supporting `JAX`, `TensorFlow`, `PyTorch` and `OpenVINO`. 
This means we won't be adding new features to `TF-Keras (Keras 2)`. 

`TensorFlow` users can benefit directly from `Keras 3`, which is included by default starting with version `2.16.1`.
 We encourage you to migrate to `Keras 3` and experience its enhanced capabilities. If you feel this specific feature is crucial for the `Keras` framework itself, please open a new issue at: https://github.com/keras-team/keras/issues

For more details and usage guide about Keras 3, refer: https://keras.io/getting_started/about/
